src/util/parallel/data.ml

```diff
@@ -0,0 +1,242 @@
+module ConcurrentHashmap =
+  functor (H: Hashtbl.HashedType) ->
+  functor (D: DefaultType) ->
+  functor (HM:Hashtbl.S with type key = H.t) ->
+  struct
+    module Bucket = ConcurrentBucket(H)(D)
+
+    type key = H.t
+    type value = D.t Atomic.t
+
+    type t = {
+      size: int Atomic.t;
+      nr_elements: int Atomic.t;
+      resize_generation: int Atomic.t;
+      buckets: Bucket.t option Atomic.t array Atomic.t;
+    }
+
+    let create () = 
+      let size = 100 in
+      {
+        size = Atomic.make size;
+        nr_elements = Atomic.make 0;
+        resize_generation = Atomic.make 0;
+        buckets = Atomic.make @@ Array.init size (fun _ -> Atomic.make None);
+      }
+
+    let to_list hm =
+      Array.fold_left (fun acc bucket ->
+          match Atomic.get bucket with
+          | None -> acc
+          | Some bucket -> acc @ Bucket.to_list bucket
+        ) [] (Atomic.get hm.buckets)
+
+    let to_seq hm =
+      let bucket_seq = Array.to_seq (Atomic.get hm.buckets) in
+      let non_atomic_bucket_seq = Seq.map Atomic.get bucket_seq in
+      let non_option_bucket_seq = Seq.filter (fun x -> x <> None) non_atomic_bucket_seq in
+      let non_atomic_bucket_seq = Seq.map (fun x -> match x with | Some x -> x | None -> failwith "This should not happen") non_option_bucket_seq in
+      Seq.flat_map Bucket.to_seq non_atomic_bucket_seq
+
+    let find_option hm key =
+      let hash = abs @@ H.hash key in
+      let bucket = Array.get (Atomic.get hm.buckets) (hash mod (Atomic.get hm.size)) in
+      match Atomic.get bucket with
+      | None -> None
+      | Some bucket -> Bucket.find_option bucket key
+
+    let find hm key =
+      match find_option hm key with
+      | None -> failwith "Key not found"
+      | Some value -> value
+
+    let mem hm key =
+      match find_option hm key with
+      | None -> false
+      | Some _ -> true
+
+
+    let rec find_create (hm : t) (key : H.t) =
+      let rec find_create_inner hm key hash buckets =
+        let bucket = Array.get buckets (hash mod (Atomic.get hm.size)) in
+        match Atomic.get bucket with
+        | None ->
+          let new_bucket = Bucket.create key in
+          let success = Atomic.compare_and_set bucket None (Some new_bucket) in
+          if success then (
+            (new_bucket.value, true) 
+          )
+          else find_create_inner hm key hash buckets 
+        | Some bucket -> 
+          let value, was_created = Bucket.find_create bucket key in
+          (value, was_created)
+      in
+      let current_generation = Atomic.get hm.resize_generation in
+      let hash = abs @@ H.hash key in
+      let value, was_created = find_create_inner hm key hash (Atomic.get hm.buckets) in
+      if (current_generation mod 2 == 0) && (Atomic.get hm.resize_generation == current_generation || not was_created) then (
+        if (Atomic.get hm.nr_elements >= Atomic.get hm.size * 2) then (
+          resize hm;
+        );
```

Comment:
If the number of elements is twice the size? Sounds odd that can happen at all...